sklearn/cluster/_hierarchical.pyx

```diff
@@ -332,3 +332,95 @@ cdef class WeightedEdge:
                                               self.weight,
                                               self.a, self.b)
 
+
+################################################################################
+# Efficient labelling/conversion of MSTs to single linkage hierarchies
+
+cdef class UnionFind(object):
+
+    cdef np.ndarray size_arr
+    cdef ITYPE_t next_label
+    cdef ITYPE_t[:] parent
+    cdef ITYPE_t *size
+
+    def __init__(self, N):
+        self.parent = -1 * np.ones(2 * N - 1, dtype=ITYPE, order='C')
+        self.next_label = N
+        self.size_arr = np.hstack((np.ones(N, dtype=ITYPE),
+                                   np.zeros(N-1, dtype=ITYPE)))
+        self.size = (<ITYPE_t *> self.size_arr.data)
```

Comment:
btw, you could similarly use a memoryview here.

Reply:
I agree ... I didn't because fewer changes seemed better, but you are quite correct.